### DISCUSSION TITLE
Does any library of utility/helpers for luasnip exist?

### DISCUSSION THREAD

**Discussion by danielo515** [L3MON4D3/LuaSnip#710] [General]
Luasnip is awesome, impressive, mind-blowing. The best snippet engine I used so far.
But with everything this powerful comes big complexity, and Luanip is quite complex.
Everything that is cool and unique to luansip requires big amounts of code you mostly copy paste without understanding much and from one file to another.
I think it will be cool to have a set of utility functions that make certain complex tasks simpler. Lodash style, not sure if you are familiar.

Regards

**Comment by atticus-sullivan** [comment]
Hm, I think what comes closest is the stuff in `luasnip.extras` (at least that I know of).
Some effort also was made in the wiki here on github (but these snippets are pretty specialized).
But having some popular reoccuring building blocks in `extras` might be a nice thing (maybe also making it easier for new users). Maybe we'll just collect some ideas and decide then if we've found things that really are commonly used (otherwise it doesn't make sense to incorporate them in this repo, just write some sort of function and use it for yourself) and can be integrated in a nice way into this repo.

**Reply by danielo515** [reply]
Integrating them here will make it easier,but having a separate library will also work fine IMO

**Reply by L3MON4D3** [reply]
+1 for extras, that's where I'd like to put useful extensions that don't have to be in the core of Luasnip, ie. stuff one might write for their own config. I'm also pretty relaxed on what is in scope there, imo it won't hurt to put in stuff that's well written (so it won't need much maintenance), even if it's only slightly useful

**Reply by L3MON4D3** [reply]
I also like the idea of a more opinionated luasnip-config repo which makes use of some newer stuff which is not explicitly implemented in here, and provides some more defaults

**Comment by L3MON4D3** [comment]
What are the big amounts of code you're referring to? Boilerplate for dynamicNode/functionNode?

**Reply by danielo515** [reply]
Yes, 90% of the times are dynamic/functioNodes, but specially choice ones.
The problem is also that, the current API is so general (it needs to be to be powerful) that is very hard to understand, and every time I have to use one of the advanced features it takes me a lot of try and error and I also have to re-visit my other snippet files.
Some of the simple approaches that are hard to understand often involve dynamic, function nodes or choice nodes or all of them 😄 

```lua
  s(
    { trig = "reqm", dscr = "Require a module but extract a method" },
    fmt("local {} = require'{}'.{}", {
      i(1, "method"),
      c(2, { t "", sn(nil, dl(nil, "user.util." .. l._1, ai[1])) }),
      rep(1),
    })
  ),
```
That is a good example of one of my snippets that, I don't even know what it does until I spent 2 or 3 minutes understanding it.
One of my main gripes (and I was about to open a separate discussion to get some help understanding that) is when to use dynamic node, snippet_node and why they are necessary.

About things that I wish they were easier:
 1. Capitalization of repeat nodes. Often I want to capitalize in some part what I'm writting in another
 2. Capitalization of what I'm typing. Will be cool to automatically get capitalized words in snippets where necessary 
 3. Sub string on repeating nodes. Like I am writing `a.b.c` then the repeat node should be c but also be b while I'm typing `a.b`
 4. Usual transformations that will help a lot are xxx-case transformation. For example, you type a library require named nvim-dev and it gets translated to nvimDev or nvim_dev or whatever you choose as casing

Of course  all those are already doable on luasnip, but often times it is so hard for me to figure out how to do that I just abandon the idea because I have work to do 😂

**Reply by L3MON4D3** [reply]
1,3 are pretty cool additions, if we can find descriptive names, we can add them to `luasnip.extras.functions` (I think we already have those)
2,4 could be achieved by a `on_leave`-event on an insertNode, so a bit more complicated, and 2 not while typing, but with this also a good fit for extras, maybe in a separate module
Conceptually, most of this `type something and have it show up differently` are pretty hard to do because dynamicNode can't depend on them self. One way out is the external update dynamic node, which can just take anything as input, for exame by opening a vim.ui.input, and inserting whatever is entered there. It's high time that one is added to extras, too.

Quickly on the differences: dynamicNode to generate nodes from text, snippetNode to group nodes together (tbh for users, that could also have just been a table, there's not much difference from outside), choiceNode to switch between nodes.

**Reply by danielo515** [reply]
for capitalizing functions what about giving them the name they are after? The ones I'm interested are Pascal case, kebab-case, CamelCase and snake_case we may put them all under a case utility ? case.pascal, case.kebab case.camel case.snake ? and make it take a node reference like the repeat node does? or add them as options to the repeat node? `rep(1,"pascal"), rep(1,"camel")`

Something similar could be added for substring, extend the existing rep node?

**Reply by danielo515** [reply]
Do you mean that 2 and 4 are already possible using the existing `on_leave` event, or that such event will be needed to support it?.
What do you think about my previous proposal?

**Reply by L3MON4D3** [reply]
Oh srry, missed your previous message.
Yes, `on_leave` already exists, see [here](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#events). #732 has some more hints, the comment I linked a good implementation.
I like the ideas :+1: Should be possible to adapt the current `rep` we provide to have this transformation+substring (maybe matching)  built-in.
We could also implement this by relying on the [transformation-syntax](https://code.visualstudio.com/docs/editor/userdefinedsnippets#_variable-transforms) provided by LSP-snippets, which we already implement anyway.

**Reply by danielo515** [reply]
You linked global events, which I don't think are useful for my use case. Maybe you meant snippet specific events? https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#snippet-data:~:text=callbacks%3A%20Contains%20functions%20that%20are%20called%20upon%20entering/leaving%20a%20node%20of%20this%20snippet.

**Reply by L3MON4D3** [reply]
Oh, yes, that's probably the better link :D
The one I shared mentions the snippet-specific callbacks only tangentially, and there's no example for it :sweat_smile:
